### PR TITLE
txnbuild allow trust (#982)

### LINKED ISSUE #982: Implement Change Trust / Allow Trust
Description here: https://www.stellar.org/developers/guides/concepts/list-of-operations.html
XDR reference here: https://github.com/stellar/stellar-core/blob/master/src/xdr/Stellar-transaction.x

- [x] Change Trust
- [x] Allow Trust

### PR DESCRIPTION
**Pull request by ire-and-curses**
Adds allow trust operation to `txnbuild`. Verified by running operation manually and inspecting returned XDR for flag change.

As usual, feel free to skip scratchpad code in `/txnbuild_examples/main.go`.

This closes #982.

### CODE REVIEW

File: exp/txnbuild/asset.go [hunk 1]
```diff
@@ -57,3 +57,25 @@ func (a *Asset) ToXDR() (xdr.Asset, error) {
...
+		return xdr.NewAllowTrustOpAsset(xdr.AssetTypeAssetTypeCreditAlphanum4, code)
+	case length >= 5 && length <= 12:
+		var code [12]byte
+		byteArray := []byte(a.Code)
+		copy(code[:], byteArray[0:length])
+		return xdr.NewAllowTrustOpAsset(xdr.AssetTypeAssetTypeCreditAlphanum12, code)
+	default:
+		return xdr.AllowTrustOpAsset{}, errors.New("Asset code length is invalid")
+	}
+}
```
Comment (bartekn) [inline exp/txnbuild/asset.go]: Yes, this belongs to `xdr/allow_trust_op_asset.go`. It already contains the other conversion (`AllowTrustOpAsset -> Asset`).

File: exp/txnbuild/asset.go [hunk 2]
```diff
@@ -57,3 +57,25 @@ func (a *Asset) ToXDR() (xdr.Asset, error) {
...
 	return xdrAsset, nil
 }
+
+// ToXDRAllowTrustOpAsset for Asset produces a corresponding XDR "allow trust" asset, used by the
+// XDR allow trust operation.
+func (a *Asset) ToXDRAllowTrustOpAsset() (xdr.AllowTrustOpAsset, error) {
+	length := len(a.Code)
+
+	// TODO: This code could be moved to XDR library as is done with xdr.Asset()
+	switch {
```
Comment (poliha) [inline exp/txnbuild/asset.go]: This code block looks familiar, if its used in other locations you might want to move this to a function

File: exp/txnbuild/allow_trust.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+type AllowTrust struct {
+	Trustor   string
+	Type      *Asset
+	Authorize bool
+	xdrOp     xdr.AllowTrustOp
+}
+
+// BuildXDR for AllowTrust returns a fully configured XDR Operation.
+func (at *AllowTrust) BuildXDR() (xdr.Operation, error) {
+	var err error
```
Comment (howardtw) [inline exp/txnbuild/allow_trust.go]: we can just do `err := ...` on line 22

File: exp/txnbuild/asset.go (hunk 1 above)
Comment (ire-and-curses) [reply exp/txnbuild/asset.go]: Raised #1013 to address this later.

File: exp/txnbuild/asset.go (hunk 2 above)
Comment (ire-and-curses) [reply exp/txnbuild/asset.go]: Yeah, there's a similar block for assets in the XDR package. Will move this later (#1013).

Comment (howardtw) [reply exp/txnbuild/asset.go]: I actually prefer just using two `if` conditions like
```
if length >= 1 && length <= 4 {
...
return xdr.NewAllowTrustOpAsset
}
if length >= 5 && length <= 12 {
...
return xdr.NewAllowTrustOpAsset
}
return ..., error

Comment (bartekn) [reply exp/txnbuild/asset.go]: It is idiomatic to use [`switch`](https://golang.org/doc/effective_go.html#switch) here:
> The expressions need not be constants or even integers, the cases are evaluated top to bottom until a match is found, and if the switch has no expression it switches on true.  It's therefore possible—and idiomatic—to write an if-else-if-else chain as a switch.

**Pull request merged**